_posts/2018-10-29-help-test-rust-2018.md

```diff
@@ -0,0 +1,131 @@
+
+## Setup: install Rust beta
+
+First things first, you'll need to install the beta release channel of Rust.
+With Rustup, it's as easy as:
+
+```console
+$ rustup install beta
+```
+
+To use this channel of Rust instead of your default, you can append a `+beta`
+to any `rustc` or cargo commands:
+
+```console
+$ rustc +beta --version
+$ cargo +beta build
+```
+
+This lets you stick to stable as the default, while using beta for your
+experiments.
+
+## Start a new project
+
+To start a new project with Rust 2018:
+
+```console
+$ cargo +beta new my-sample-project
+```
+
+Nothing changes! Well, something changed. Check out `Cargo.toml`:
+
+```toml
+[package]
+name = "my-sample-project"
+version = "0.1.0"
+authors = ["Your Name <you@example.com>"]
+edition = "2018"
+
+[dependencies]
+```
+
+That new `edition = "2018"` key/value pair means you're working with Rust 2018.
+If it doesn't exist, it's the same as `edition = "2015"`, so all
+existing projects keep working.
+
+## Convert an existing project
+
+You can also convert an existing project to Rust 2018. Remember, none of your
+dependencies need to be updated for this to work; Rust 2018 and 2015
+interoperate seamlessly!
+
+The first step is to run `cargo fix`:
+
+```console
+$ cargo fix --edition
+```
+
+This will check your code, and automatically fix any issues that it can.
+`cargo fix` is still pretty new, and so it can't always fix your code
+automatically. If `cargo fix` can't fix something, it will print the warning
+that it cannot fix to the console. If you see one of these warnings, you'll
+have to update your code manually. See the corresponding section of the
+edition guide for help, and if you have problems, please seek help at the
+user's forums.
+
+Keep running `cargo fix --edition` until you have no more warnings.
+
+Congrats! Your code is now valid in both Rust 2015 and Rust 2018!
+
+Once this is done, you can commit to Rust 2018 by updating
+your `Cargo.toml`:
+
+```toml
+[package]
+name = "my-sample-project"
+version = "0.1.0"
+authors = ["Your Name <you@example.com>"]
+edition = "2018"
+
+[dependencies]
+```
+
+See that `edition = "2018"`? That's what opts you in to the new features.
+Set it, `cargo +beta build`, and you should be good to go!
+
+## Writing idiomatic code
+
+We have some new lints that suggest using certain idioms in your Rust 2018
+code. We don't have them turned on by default yet. To see what your code would
+look like, you can use `cargo fix`:
+
+```console
+$ cargo +beta fix --edition-idioms
+```
+
+If things look great, or things look terrible, please let us know! We hope to make
+these lints warn by default after gaining some experience with them and working
+out the bugs.
+
+> The `--edition-idioms` flag applies only to the "current crate"; if you want
+> to run it against a workspace is necessary to use a workaround with `RUSTFLAGS`
+> in order to execute it in all the workspace members.
+> 
+>     $ RUSTFLAGS='-Wrust_2018_idioms' cargo fix --all
```

Comment:
I'm not sure we should call out idiom lints at all. We've punted on multiple issues with them IIRC and I don't believe they're ready for a public announcement like this.

Reply:
Hm, so the point of this is to get feedback from a wide variety of people; wouldn't we *want* to mention them in order to see how they're currently doing?

Reply:
I believe some of these might currently be in "eat your socks" mode which makes me worry that the perspective of new people will be "editions are painful to transition to, the lines work poorly." On the other hand not saying anything might lead to that being a common question... I think I'm leaning towards neutrality on this - that is, whatever you think is best works for me.

Reply:
Maybe @centril has some opinions?

If they are that bad, I'm okay with not doing it.

Reply:
@Mark-Simulacrum @steveklabnik I can see a case for being very conservative about recommendations for the final stable edition release, but I think we should definitely ask people to test the idiom lints _when we're explicitly asking people to test things_! Note that the most problematic lints were already demoted from the group in rust-lang/rust#52926.

My take on the suriviors follows. (Full disclosure: I implemented `ellipsis_inclusive_range_patterns`, `elided_lifetimes_in_paths`, and `explicit_outlives_requirements`: the reader must decide for themself whether that makes me knowledgable or biased.)

 * `bare_trait_objects`: no problems that I'm aware of?

 * `unused_extern_crates`: false-positives reported in rust-lang/rust#53964 (someone should take a look at this)

 * `ellipsis_inclusive_range_patterns`: the suggestion is [currently marked as not-auto-fixable](https://github.com/rust-lang/rust/issues/51043#issuecomment-404390798), but the blocking issue there has been resolved—it's likely I could make it machine-applicable this week

 * `elided_lifetimes_in_paths`: no problems that I'm aware of?

 * `explicit_outlives_requirements`: some false-negatives (rust-lang/rust#54630), and rust-lang/rust#53934 is a blocker for adding multi-span suggestion support to rustfix